src/control/lib/hardware/topology.go

```diff
@@ -34,20 +41,130 @@ func (t *Topology) AllDevices() map[string]*Device {
 	return devsByName
 }
 
+func PrintTopology(t *Topology, output io.Writer) error {
+	if t == nil {
+		fmt.Fprintf(output, "No topology information available\n")
+	}
+
+	for _, numaNode := range t.NUMANodes {
+		coreSet := &system.RankSet{}
+		for _, core := range numaNode.Cores {
+			coreSet.Add(system.Rank(core.ID))
+		}
+
+		fmt.Fprintf(output, "NUMA node %d (%s)\n", numaNode.ID, numaNode.Set)
+		fmt.Fprintf(output, "  CPU cores: %s\n", coreSet)
+		fmt.Fprintf(output, "  PCI devices: \n")
+		for addr, devs := range numaNode.PCIDevices {
+			fmt.Fprintf(output, "    %s\n", &addr)
+			for _, dev := range devs {
+				fmt.Fprintf(output, "      %s\n", dev)
+			}
+		}
+		fmt.Fprintf(output, "  PCI buses: \n")
+		for _, bus := range numaNode.PCIBuses {
+			fmt.Fprintf(output, "    %s\n", bus)
+		}
+	}
+
+	return nil
+}
+
 type (
+	CPUCore struct {
+		ID       uint      `json:"id"`
+		NUMANode *NUMANode `json:"-"`
+	}
+
+	// PCIDevice represents an individual hardware device.
+	PCIDevice struct {
+		Name      string            `json:"name"`
+		Type      DeviceType        `json:"type"`
+		NUMANode  *NUMANode         `json:"-"`
+		PCIAddr   common.PCIAddress `json:"pci_address"`
+		LinkSpeed float64           `json:"link_speed"`
+	}
+
+	PCIBus struct {
+		LowAddress  common.PCIAddress `json:"low_address"`
+		HighAddress common.PCIAddress `json:"high_address"`
+		NUMANode    *NUMANode         `json:"-"`
+	}
+
 	// NUMANode represents an individual NUMA node in the system and the devices associated with it.
 	NUMANode struct {
-		ID       uint       `json:"id"`
-		NumCores uint       `json:"num_cores"`
-		Devices  PCIDevices `json:"devices"`
+		ID         uint       `json:"id"`
+		Set        string     `json:"set"`
+		Cores      []CPUCore  `json:"cores"`
+		PCIBuses   []*PCIBus  `json:"pci_buses"`
```

Comment:
Might we want the PCI buses to be in a map as well? If it is something we will ever want to search by bus ID

Reply:
That I'm not sure about. My preference would be to keep it simple for now and if we have a need to optimize it we can add it later.